### PR TITLE
feat(contracts): implement semaphore gatekeeper

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

Implement a MACI gatekeeper that uses semaphore to gatekeep access to group members only

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](https://maci.pse.dev/docs/contributing) and [code of conduct](https://maci.pse.dev/docs/contributing/code-of-conduct).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847).
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: contracts/contracts/gatekeepers/SemaphoreGatekeeper.sol [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    // register the nullifier so it cannot be called again with the same one
+    registeredIdentities[proof.nullifier] = true;
+
+    /// @notice force the scope to be the group id
+    /// this way we can ensure that the nullifier is
+    /// hash(secret, groupId)
+    /// and thus one identity can only be registered once via this
+    /// gatekeeper
+    proof.scope = groupId;
```
Comment (cedoor) [inline contracts/contracts/gatekeepers/SemaphoreGatekeeper.sol:78]: This should be equivalent to checking that `scope` equals `groupId`, which could be easier to read.

File: contracts/contracts/mocks/MockSemaphore.sol [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import { ISemaphore } from "../interfaces/ISemaphore.sol";
+
+/// @title MockSemaphore
+/// @notice A mock contract to test the Semaphore gatekeeper
+contract MockSemaphore {
+  /// @notice The group id
+  uint256 public groupId;
```
Comment (0xmad) [inline contracts/contracts/mocks/MockSemaphore.sol:10]: Let's make it immutable

File: contracts/contracts/mocks/MockSemaphore.sol [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import { ISemaphore } from "../interfaces/ISemaphore.sol";
+
+/// @title MockSemaphore
+/// @notice A mock contract to test the Semaphore gatekeeper
+contract MockSemaphore {
```
Comment (0xmad) [inline contracts/contracts/mocks/MockSemaphore.sol:8]: Maybe it makes sense to inherit from ISemaphore?

**Review by cedoor — CHANGES_REQUESTED** [review]
Looks good overall, I just have a question. Shouldn't there be a mapping to save the id commitments? I can only see one where you're storing the nullifiers.

**Review by 0xmad — COMMENTED** [review]
@ctrlc03 thanks, just few suggestions

File: contracts/contracts/gatekeepers/SemaphoreGatekeeper.sol (hunk 1 above)
Comment (ctrlc03) [reply contracts/contracts/gatekeepers/SemaphoreGatekeeper.sol]: good shout, I just simplified the code

**Review by 0xmad — APPROVED** [review]
@ctrlc03 thanks!

**Review by kittybest — APPROVED** [review]
thanks!

**Review by cedoor — APPROVED** [review]
LGTM!

**Review by crisgarner — APPROVED** [review]

**Pull request merged**